### ISSUE TITLE
Requesting YUV Support for image.frombuffer()

### ISSUE THREAD

**Issue by roninpawn** [pygame/pygame#2967]
**Description**

I know PyGame deprecated its YUV overlay support, and that's all well and good!

This request is for PyGame's [image.frombuffer()](https://www.pygame.org/docs/ref/image.html#pygame.image.frombuffer) method to support the ingest of a 'YUV420p' bytestream, and that alone!

See, I've just pushed my first release of [a Python script ](https://github.com/roninpawn/ffmpeg_videostream)that allows raw video frame data to be pulled into Python using FFmpeg and FFprobe. With the latest build of FFmpeg on Windows I'm reading in ~220 frames per second from your average 1080p .mp4, which I can then blit to screen via a stock-standard PyGame window at over 80fps.

But to do so, I have to invoke two extra lines.

```python
arr = np.frombuffer(frame, np.uint8).reshape(video.shape()[1]*3//2, video.shape()[0])
img = cv2.cvtColor(arr, cv2.COLOR_YUV2BGR_I420)
```

Before I can hand the frame to PyGame, the raw bytestream has to be reshaped using numpy, _(zero slowdown there)_ so I can hand it over to OpenCV for conversion into the BGR pixel space that PyGame supports. _(lots of slowdown there)_ Of course, no matter where the YUV conversion takes place, it's going to cost time, but I was simply thinking, "wouldn't it be lovely," if I could pass the YUV input straight into `image.frombuffer()` the same way I can when the source is BGR.

```python
img = pygame.image.frombuffer(img, video.shape(), "BGR")
# except...
img = pygame.image.frombuffer(img, video.shape(), "YUV")
```

Now, I see how this could open up a Pandora's box of varying formats to support, but I think it would be fair to onboard this particular YUV standard and handwave everything else away. The interlaced, YUV420p, 12 bit-per-pixel format is THE standard for video. DVDs, Blu-rays, MP4s, WEBMs... It's all 12-bit YUV 4:2:0 encoding, because it's the smallest you can pack the pixels before the human eye notices a loss. And whether its YouTube minimizing bandwidth or Disney trying to cram three hours on a single side of a disc, it's all packaged up as YUV420p.

PyGame's `.frombuffer()` being able to ingest those frames raw and scoop them onto screen without invoking interstitial calls to Numpy and OpenCV would be a boon for Python's ability to render software video. And I would expect to find a lot of people coming up with new and interesting implementations for that in PyGame environments.

So that's my request. Thanks!

**Comment by MyreMylar** [comment]
Don't think there is that much improvement to be made via the SDL/pygame software renderer, everything there has to eventually be in an RGB type format for blitting as that is what the SDL surfaces for the displays/windows are.  Are you currently doing a YUV -> BGR then a BGR -> RGB ? I guess that could potentially have a step removed.

However, also potentially of interest is the new stuff using the SDL2 renderer (most of this is currently under `_sdl2`) that could likely do something. See:

https://www.programmerall.com/article/8711762643/
https://wiki.libsdl.org/SDL_UpdateYUVTexture

Basically you can use graphics card hardware via OpenGL to directly shove YUV textures onto the screen.

Would likely have to have an interface something like `pygame.texture.frombuffer()` to return a SDL Texture like wrapper instead of a pygame surface.

Not sure if that is of interest as you would lose the easy direct pixel access of software rendering.

**Comment by Starbuck5** [comment]
It may be faster to use opencv to convert from YUV to RGB directly, that may lead to a faster blit on the display surface.

Fun fact, there is a YUV to RGB conversion function in the pygame codebase. It's in the camera module and is only available internally to convert YUV frames into RGB surfaces.

Also, there are YUV surfaces in SDL. https://wiki.libsdl.org/SDL_PixelFormatEnum.

> https://wiki.libsdl.org/SDL_UpdateYUVTexture

I believe this function is here for if the data is in multiple planes, but it has the same utility and handicaps as SDL_UpdateTexture. Like the Texture needs to be in the right format. Which varies from render backend to render backend, and I don't think SDL supports many Texture formats on each backend.

**Comment by Starbuck5** [comment]
I wonder what would happen if you passed YUV data and the YUV Pixelformat to `SDL_CreateRGBSurfaceWithFormatFrom`.

**Comment by roninpawn** [comment]
I suppose it's fair enough, as I consider it now, not to look for PyGame to create or maintain an access method - which basically comes down to a color/pixel space conversion method - for this one YUV format. After all, wherever there exists, or comes to exist, a better mouse trap (read: more efficient matrix / hardware-enabled conversion), it should be up to the developer to implement it as necessary. That said, it would be a nice feature and where there might be something better to be found, the developer could always circumvent the PyGame method with their own and go straight to an RGB / BGR method.

## YUV > BGR > RGB
After more testing, all I can say is that my results are a bit mixed. I've now seen scenarios where it appears that RGB is decidedly faster. Where BGR is decidedly faster. And where the difference in results are so negligible it's hard to call. But if I boil down my experiences running gobs and gobs of tests during development, and mix it with the latest tests I whipped up to try RGB vs BGR, it seems that:

- RGB is slightly faster when rendering an unmodified frame.
- BGR is slightly faster when invoking a pygame.transform() on the input frame.

My test bed throughout development creates a 1600 x 900 PyGame window, pulls the source video in from FFmpeg at half its original size, then pygame.transform.scale()'s that frame to be 0.75x the size of the window, (900 x 675) blits the current fps onto the frame 10x10 pixels in from the top, left corner, then draws the frame into the window at an offset of x=10, y=10, updating only the necessary rect[] of pixels.

I created this test bed to work a little of everything at once. And I've, of course, run other, 'purer' tests. It's hard to get consistent results with CPU rendering, too. If I open a tab on a browser, a chunk of CPU will be dedicated to that tab for a bit, even once the browser window's off screen. So take me with a grain of salt.

But my results seem to say that BGR is faster across transform operations, RGB is faster with an untouched 1080p frame, and in middling-resolutions with an untouched frame, there's virtually no difference. Which makes sense considering all of the different types of potentially-blocking processes involved. Once a processing block becomes less than a given I/O block, or vice versa, it would cease to make a difference in the time it takes to get the frame to screen.

## SDL2
I'm glad you both brought it up, because the next rabbit hole was definitely going to be finding hardware-accelerated methods for pushing the YUV frames to screen; and even just for converting from the YUV to an RGB / BGR color space. And I'm going to have a good look at Python's SDL libraries soon.

Thanks for hearing me out and for the helpful replies!

**Comment by Starbuck5** [comment]
I'm having difficulties understanding the actual YUV format here.

These are our options from SDL:
<html>
<body>


SDL_PIXELFORMAT_YV12 | planar mode: Y + V + U (3 planes)
-- | --
SDL_PIXELFORMAT_IYUV | planar mode: Y + U + V (3 planes)
SDL_PIXELFORMAT_YUY2 | packed mode: Y0+U0+Y1+V0 (1 plane)
SDL_PIXELFORMAT_UYVY | packed mode: U0+Y0+V0+Y1 (1 plane)
SDL_PIXELFORMAT_YVYU | packed mode: Y0+V0+Y1+U0 (1 plane)
SDL_PIXELFORMAT_NV12 | planar mode: Y + U/V interleaved (2 planes) (>= SDL 2.0.4)
SDL_PIXELFORMAT_NV21 | planar mode: Y + V/U interleaved (2 planes) (>= SDL 2.0.4)


</body>
</html>

Theoretically, we could try making a surface with YUV data directly, and then SDL could handle the conversion when it's blitted.

I don't know if that would be more efficient.

**Comment by roninpawn** [comment]
From the table, it looks like the 'IYUV' option would be the one for unpacking I420. By 3 planes, its talking about three discreet chunks of data that don't overlap. So the ENTIRE 'Y' channel (greyscale) is stored in one contiguous 'plane;' then the entire 'U'  channel (blue); then the entire 'V' channel. (red)

YUV 4:2:0 (which is Y'CbCr, I think?) looks like the image below when decoded as though it's RGB data. The 'U' and 'V' channels that carry the color information are tacked onto the end at 1/4 the size of the full grayscale 'Y' frame. So you end up getting this 'two-rows-per-row,' effect if you force U and V into the shape of the 'Y' channel's horizontal resolution.

![YUV 4:2:0](https://user-images.githubusercontent.com/49294050/149018977-b6cfcb30-fea5-4ed1-ab58-898976bcffb4.png)

Of the options, 'SDL_PIXELFORMAT_IYUV' describes itself as 3 planes arranged in the typical Y, followed by U, followed by V format... So, I figure that's the one.

**Comment by Starbuck5** [comment]
What if you tell ffmpeg-python to read it in as RGB32? It's going to need to be translated into RGB eventually, probably. So that seems like an easy place to do it. https://github.com/roninpawn/ffmpeg_videostream/blob/master/ffmpeg_videostream.py#L135

**Comment by roninpawn** [comment]
> What if you tell ffmpeg-python to read it in as RGB32? It's going to need to be translated into RGB eventually, probably. So that seems like an easy place to do it. https://github.com/roninpawn/ffmpeg_videostream/blob/master/ffmpeg_videostream.py#L135

I whipped up a basic test bed to show my work on this.

TLDR:

```       Begin Pygame DRAW Tests
        --- TEST #5: FFmpeg > RGB24 > PyGame
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 110.814 seconds.
        Effective rate of 74.1 frames per second.

        --- TEST #6: FFmpeg > YUV420p > OpenCV:RGB > PyGame
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 80.303 seconds.
        Effective rate of 102.3 frames per second.
```

When drawing the frames to a PyGame window, processing YUV420p to RGB using OpenCV is ~38% faster than having FFmpeg output the bytes as RGB24. When reading the frames in RAW and discarding them, its +120% faster to get YUV than RGB. And after converting YUV to RGB in memory (but not drawing) it's +60% faster.

Running this latest batch of tests also confirms that sending RGB to PyGame is slightly faster than BGR. However, when a scale transform is involved, the results become mixed and inconclusive.

I've added the simple test script I wrote to the repo, here:
https://github.com/roninpawn/ffmpeg_videostream/blob/master/tests/pygame_tests.py

And below is a transcript of my results across both a 1080p and 720p mp4.
Tested on an AMD Ryzen 5 3600- 6 core (12 virtual) on Windows 10 w/ 16Gb's RAM. (timings forgotten)
```
    1920x1080:
        --- TEST #1: FFmpeg > RGB24 > RAW (no draw)
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 85.251 seconds.
        Effective rate of 96.4 frames per second.

        --- TEST #2: FFmpeg > YUV420p > RAW (no draw)
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 38.444 seconds.
        Effective rate of 213.7 frames per second.

        --- TEST #3: FFmpeg > YUV420p > OpenCV:RGB (no draw)
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 52.765 seconds.
        Effective rate of 155.7 frames per second.

        --- TEST #4: FFmpeg > YUV420p > OpenCV:BGR (no draw)
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 54.155 seconds.
        Effective rate of 151.7 frames per second.

        Begin Pygame DRAW Tests
        --- TEST #5: FFmpeg > RGB24 > PyGame
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 110.814 seconds.
        Effective rate of 74.1 frames per second.

        --- TEST #6: FFmpeg > YUV420p > OpenCV:RGB > PyGame
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 80.303 seconds.
        Effective rate of 102.3 frames per second.

        --- TEST #7: FFmpeg > YUV420p > OpenCV:BGR > PyGame
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 82.463 seconds.
        Effective rate of 99.6 frames per second.

        --- TEST #8: PyGame scales frame in BGR
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 89.591 seconds.
        Effective rate of 91.7 frames per second.

        --- TEST #9: PyGame scales frame in RGB
        Processed 8216 frames at (1920, 1080) resolution from 'test_video.mp4' in 88.831 seconds.
        Effective rate of 92.5 frames per second.

    1280x720:
        --- TEST #1: FFmpeg > RGB24 > RAW (no draw)
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 15.701 seconds.
        Effective rate of 243.2 frames per second.

        --- TEST #2: FFmpeg > YUV420p > RAW (no draw)
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 6.753 seconds.
        Effective rate of 565.5 frames per second.

        --- TEST #3: FFmpeg > YUV420p > OpenCV:RGB (no draw)
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 10.635 seconds.
        Effective rate of 359.1 frames per second.

        --- TEST #4: FFmpeg > YUV420p > OpenCV:BGR (no draw)
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 10.553 seconds.
        Effective rate of 361.9 frames per second.

        Begin Pygame DRAW Tests
        --- TEST #5: FFmpeg > RGB24 > PyGame
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 21.44 seconds.
        Effective rate of 178.1 frames per second.

        --- TEST #6: FFmpeg > YUV420p > OpenCV:RGB > PyGame
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 15.994 seconds.
        Effective rate of 238.8 frames per second.

        --- TEST #7: FFmpeg > YUV420p > OpenCV:BGR > PyGame
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 16.04 seconds.
        Effective rate of 238.1 frames per second.

        --- TEST #8: PyGame scales frame in BGR
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 17.386 seconds.
        Effective rate of 219.7 frames per second.

        --- TEST #9: PyGame scales frame in RGB
        Processed 3819 frames at (1280, 720) resolution from '720_test.mp4' in 17.493 seconds.
        Effective rate of 218.3 frames per second.
```
